atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/VersionedEventStore.java

```diff
@@ -63,26 +64,36 @@ LockWatchEvents retentionEvents(Optional<Long> earliestSequenceToKeep) {
             return LockWatchEvents.builder().build();
         }
 
+        // Guarantees that we remove some events while still also potentially performing further retention
         if (eventMap.size() > maxEvents) {
-            return retentionEvents(eventMap.size() - maxEvents, Long.MAX_VALUE);
+            List<LockWatchEvent> overMaxSizeEvents = retentionEvents(eventMap.size() - maxEvents, Long.MAX_VALUE);
+            List<LockWatchEvent> restOfEvents =
```

Comment:
This is kinda difficult to parse because it's not immediately obvious that the method call will modify `eventMap`, but can't think of a way to make it better...

Reply:
Commented on this.